embassy-rp/src/clocks.rs

```
@@ -783,45 +1409,99 @@ fn start_xosc(crystal_hz: u32, delay_multiplier: u32) {
     assert!(vco_freq >= 750_000_000 && vco_freq <= 1_800_000_000);
 
-    // Load VCO-related dividers before starting VCO
+    // We follow the SDK's approach to PLL configuration which is:
+    // 1. Power down PLL
+    // 2. Configure the reference divider
+    // 3. Configure the feedback divider
+    // 4. Power up PLL and VCO
+    // 5. Wait for PLL to lock
+    // 6. Configure post-dividers
+    // 7. Enable post-divider output
+
+    // 1. Power down PLL before configuration
+    p.pwr().write(|w| {
+        w.set_pd(true); // Power down the PLL
+        w.set_vcopd(true); // Power down the VCO
+        w.set_postdivpd(true); // Power down the post divider
+        w.set_dsmpd(true); // Disable fractional mode
+        *w
+    });
+
+    // Short delay after powering down
+    cortex_m::asm::delay(10);
+
+    // 2. Configure reference divider first
     p.cs().write(|w| w.set_refdiv(config.refdiv as _));
+
+    // 3. Configure feedback divider
     p.fbdiv_int().write(|w| w.set_fbdiv_int(config.fbdiv));
 
-    // Turn on PLL
-    let pwr = p.pwr().write(|w| {
-        w.set_dsmpd(true); // "nothing is achieved by setting this low"
-        w.set_pd(false);
-        w.set_vcopd(false);
-        w.set_postdivpd(true);
+    // 4. Power up PLL and VCO, but keep post divider powered down during initial lock
+    p.pwr().write(|w| {
+        w.set_pd(false); // Power up the PLL
+        w.set_vcopd(false); // Power up the VCO
+        w.set_postdivpd(true); // Keep post divider powered down during initial lock
+        w.set_dsmpd(true); // Disable fractional mode (simpler configuration)
         *w
     });
 
-    // Wait for PLL to lock
-    while !p.cs().read().lock() {}
+    // 5. Wait for PLL to lock with a timeout
+    let mut timeout = 1_000_000; // Reasonable timeout value
+    while !p.cs().read().lock() {
+        timeout -= 1;
+        if timeout == 0 {
+            // PLL failed to lock, return 0 to indicate failure
+            return 0;
```

You might want to return an Err() here instead of expecting a user to know that 0 is an error.

---

Not sure if I agree, or more precisely: What good can we do here? I changed this to return Err(), but the only consumer is the init() fn, so my best idea was to use the newly minted result, but then just panic on Err. Is that what You had in mind? Are we better off that way? Could we not just panic here instead of retunrning 0? After all, nothing in the firmware will work anyway, we can just as well abort early?

---

Returning Err instead of a magic number than means error allows you to follow standard Rust error handling patterns.

As to "should we just panic": no. This isn't an unrecoverable error.
Communicating that information is important, even if you're just panicking at the next level at the moment.

---

Okay, got it i think. So basically returning Err makes the fn more useful in case the surrounding system ever needs it for something else or can someday recover from an error here.